src/rust/index.rs

```diff
@@ -0,0 +1,413 @@
+        drop(minhash_guard);
+
+        // Process k-mers for this protein
+        let protein_kmers = self.process_protein_kmers(&uniprot_entry.sequence, &small_sig)?;
+
+        let protein = Protein::new(uniprot_entry, protein_kmers);
+
+        // Update global structures
+        {
+            let mut collection = self.collection.lock().unwrap();
+            let sigs = vec![sig];
+            let new_collection = Collection::from_sigs(sigs)?;
+            *collection = new_collection;
+
+            let mut proteins = self.protein_info.lock().unwrap();
+            proteins.insert(md5sum.clone(), protein.clone());
+        }
+
+        // Store in RocksDB
+        self.db.put(format!("protein:{}", md5sum).as_bytes(), bincode::serialize(&protein)?)?;
+
+        Ok(())
+    }
+
+    pub fn process_protein_kmers(
+        &self,
+        sequence: &str,
+        signature: &SmallSignature,
+    ) -> Result<KmerSignature> {
+        let minhash_guard = signature.minhash.clone();
+        let ksize = minhash_guard.ksize() as usize;
+        let seed = minhash_guard.seed();
+        let hashvals = minhash_guard.to_vec();
+        drop(minhash_guard);
+
+        let mut signature_kmers = KmerSignature::new(SmallSignature {
+            location: signature.location.clone(),
+            name: signature.name.clone(),
+            md5sum: signature.md5sum.clone(),
+            minhash: signature.minhash.clone(),
+        });
+
+        for i in 0..sequence.len().saturating_sub(ksize - 1) {
+            let kmer = &sequence[i..i + ksize];
+
+            // Process the k-mer to get encoded version
+            let (encoded_kmer, original_kmer) = self.encode_kmer(kmer);
+
+            // Get the hash from the minhash implementation
+            let hashval = _hash_murmur(encoded_kmer.as_bytes(), seed);
+
+            // If this hashval is in the minhash, then save its k-mer positions
+            if !hashvals.contains(&hashval) {
+                let kmer_info =
+                    signature_kmers.kmer_infos.entry(hashval).or_insert_with(|| KmerInfo {
+                        hashval,
+                        encoded_kmer: encoded_kmer.clone(),
+                        original_kmer_to_position: HashMap::new(),
+                    });
+
+                kmer_info
+                    .original_kmer_to_position
+                    .entry(original_kmer)
+                    .or_insert_with(Vec::new)
+                    .push(i);
+            }
+        }
+
+        Ok(signature_kmers)
+    }
+
+    /// Process a k-mer to get its encoded version
+    pub fn encode_kmer(&self, kmer: &str) -> (String, String) {
+        // Create the encoded k-mer by applying the encoding function
+        let encoded: Vec<u8> = kmer.chars().map(|c| (self.encoding_fn)(c as u8)).collect();
+        let encoded_str = str::from_utf8(&encoded).unwrap();
+
+        (encoded_str.to_owned(), kmer.to_owned())
+    }
```

Comment:
```suggestion
    pub fn encode_kmer(&self, kmer: &str) -> Result<(String, String)> {
        let encoded: Vec<u8> = kmer.as_bytes().iter().map(|&b| (self.encoding_fn)(b)).collect();
        let encoded_str = std::str::from_utf8(&encoded)?;
        Ok((encoded_str.to_owned(), kmer.to_owned()))
    }
```
Return a result for better error propogation.

I was tempted to use a lifetime on `kmer`, but it has to be `to_string()` anyway when it i used in `kmer_info`

**NOTE**: this instigated changes when this was called in the function above and the tests